zkevm-circuits/src/evm_circuit/util/math_gadget/addwords.rs

```diff
@@ -0,0 +1,294 @@
+            }
+            let sum = input_words[N_ADDENDS];
+            self.sum.assign(region, offset, Some(sum.to_le_bytes()))?;
+
+            let addends = input_words[0..N_ADDENDS].try_into().unwrap();
+            self.addwords_gadget.assign(region, 0, addends, sum)?;
+
+            Ok(())
+        }
+    }
+
+    #[test]
+    fn test_add_0_0() {
+        test_math_gadget_container::<Fr, AddWordsTestContainer<Fr, 2, true>>(
+            vec![Word::from(0), Word::from(0), Word::from(0)],
+            true,
+        );
+    }
+
+    #[test]
+    fn test_addwords_1_1() {
+        test_math_gadget_container::<Fr, AddWordsTestContainer<Fr, 2, true>>(
+            vec![Word::from(1), Word::from(1), Word::from(2)],
+            true,
+        );
+    }
+
+    #[test]
+    fn test_addwords_1000_1000() {
+        test_math_gadget_container::<Fr, AddWordsTestContainer<Fr, 2, true>>(
+            vec![Word::from(1000), Word::from(1000), Word::from(2000)],
+            true,
+        );
+    }
+
+    #[test]
+    fn test_addwords_to_wordmax() {
+        test_math_gadget_container::<Fr, AddWordsTestContainer<Fr, 2, true>>(
+            vec![Word::MAX - 1, Word::from(1), Word::MAX],
+            true,
+        );
+    }
+
+    #[test]
+    fn test_addwords_overflow() {
+        test_math_gadget_container::<Fr, AddWordsTestContainer<Fr, 2, true>>(
+            vec![Word::MAX, Word::from(1), Word::from(0)],
+            false,
+        );
+    }
+
+    #[test]
+    fn test_addwords_neq0() {
+        test_math_gadget_container::<Fr, AddWordsTestContainer<Fr, 2, true>>(
+            vec![Word::from(1), Word::from(0), Word::from(0)],
+            false,
+        );
+    }
+
+    #[test]
+    fn test_addwords_neq2() {
+        test_math_gadget_container::<Fr, AddWordsTestContainer<Fr, 2, true>>(
+            vec![Word::from(2), Word::from(1), Word::from(2)],
+            false,
+        );
+    }
+
+    //non overflow check cases
+    #[test]
+    fn test_addwords_no_overflow_check() {
+        test_math_gadget_container::<Fr, AddWordsTestContainer<Fr, 2, false>>(
+            vec![Word::MAX, Word::from(1), Word::from(0)],
+            true,
+        );
+    }
+
+    #[test]
+    fn test_addwords_3_addends() {
+        test_math_gadget_container::<Fr, AddWordsTestContainer<Fr, 3, false>>(
+            vec![Word::from(0), Word::from(1), Word::from(0), Word::from(1)],
+            true,
+        );
+    }
+
+    #[test]
+    fn test_addwords_3_addends_with_overflow_check() {
+        test_math_gadget_container::<Fr, AddWordsTestContainer<Fr, 3, true>>(
+            vec![Word::MAX, Word::from(1), Word::from(0), Word::from(0)],
+            false,
+        );
+    }
+
+    #[test]
+    fn test_addwords_3_addends_without_overflow_check() {
+        test_math_gadget_container::<Fr, AddWordsTestContainer<Fr, 3, false>>(
+            vec![Word::MAX, Word::from(1), Word::from(0), Word::from(0)],
+            true,
+        );
+    }
```

Comment:
`MAX_HI`/`MAX_LO` testing would be great here as well.
Also `carry(&self)` at least (maybe some of the others as well) are used externally so ideally tested here.

Reply:
Fixed